File: pset6/readability/readability.py
from cs50 import get_string


def main():
    text = get_string("Text: ")
    grade = readibility(text)
    if grade > 15:
        print("Grade 16+")
    elif grade < 1:
        print("Before Grade 1")
    else:
        print(f"Grade {grade}")


# readibility determines the grade level of the text.
def readibility(text):

    # define vars
    letters = 0
    words = 1
    sentences = 0

    # count
    for i in range(len(text)):
        ch = text[i]

        # add
        if ch.isalpha():
            letters += 1
        elif ch == " ":
            words += 1
        elif ch in [".", "!", "?"]:
            sentences += 1

    # apply formula
    lx = float(letters) / float(words) * 100.0
    sx = float(sentences) / float(words) * 100.0
    return round(0.0588 * lx - 0.296 * sx - 15.8)


main()

File: pset2/readability/readability.c
# include <stdio.h>
# include <cs50.h>
# include <string.h>
# include <ctype.h>
# include <math.h>

int letters(string s);
int words(string s);
int sentences(string s);
int coleman_liau(string s);

int main(void)
{
    // get the grade
    string text = get_string("Text: ");
    int grade = coleman_liau(text);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

// coleman_liau calculates the index of the readibility
int coleman_liau(string s)
{
    // get data
    int ls = letters(s);
    int ws = words(s);
    int ss = sentences(s);
    int len = strlen(s);

    // calculates letters per 100 words
    float lx = (float) ls / (float) ws * 100;
    // callculates sentences per 100 words;
    float sx = (float) ss / (float) ws * 100;

    // apply the formula
    double index = 0.0588 * lx - 0.296 * sx - 15.8;
    int result = (int) round(index);

    return result;
}

// letters calculates the number of letters in the text.
int letters(string s)
{
    int count = 0;

    // range over the string
    for (int i = 0, l = strlen(s); i < l; i++)
    {
        if (isalpha(s[i]))
        {
            count++;
        }
    }

    return count;
}

// words counts the number of the words in the text.
int words(string s)
{
    int count = 0;

    // range over the string
    for (int i = 0, l = strlen(s); i < l; i++)
    {
        if (s[i] == ' ')
        {
            count++;
        }
    }

    return count + 1;
}

// sentences counts the number of the sentences in the text.
int sentences(string s)
{
    int count = 0;

    // range over the string
    for (int i = 0, l = strlen(s); i < l; i++)
    {
        if (s[i] == '.' || s[i] == '!' || s[i] == '?')
        {
            count++;
        }
    }

    return count;
}